x-pack/plugins/lens/public/indexpattern_datasource/counter_rate/index.ts

```diff
@@ -0,0 +1,141 @@
+ *   will be set to `undefined`.
+ * * If the row value contains `null` or `undefined`, it will be ignored and the output cell will be set to `undefined`
+ * * If the value of the previous row of the same series contains `null` or `undefined`, the output cell of the current row will be set to `undefined` as well
+ * * For all values besides `null` and `undefined`, the value will be cast to a number before it's used in the
+ *   calculation of the current series even if this results in `NaN` (like in case of objects).
+ * * To determine separate series defined by the `by` columns, the values of these columns will be cast to strings
+ *   before comparison. If the values are objects, the return value of their `toString` method will be used for comparison.
+ *   Missing values (`null` and `undefined`) will be treated as empty strings.
+ */
+export const counterRate: ExpressionFunctionCounterRate = {
+  name: 'lens_counter_rate',
+  type: 'datatable',
+
+  inputTypes: ['datatable'],
+
+  help: i18n.translate('xpack.lens.functions.counterRate.help', {
+    defaultMessage: 'Calculates the counter rate of a column in a data table',
+  }),
+
+  args: {
+    by: {
+      help: i18n.translate('xpack.lens.functions.counterRate.args.byHelpText', {
+        defaultMessage: 'Column to split the counter rate calculation by',
+      }),
+      multi: true,
+      types: ['string'],
+      required: false,
+    },
+    inputColumnId: {
+      help: i18n.translate('xpack.lens.functions.counterRate.args.inputColumnIdHelpText', {
+        defaultMessage: 'Column to calculate the counter rate of',
+      }),
+      types: ['string'],
+      required: true,
+    },
+    outputColumnId: {
+      help: i18n.translate('xpack.lens.functions.counterRate.args.outputColumnIdHelpText', {
+        defaultMessage: 'Column to store the resulting counter rate in',
+      }),
+      types: ['string'],
+      required: true,
+    },
+    outputColumnName: {
+      help: i18n.translate('xpack.lens.functions.counterRate.args.outputColumnNameHelpText', {
+        defaultMessage: 'Name of the column to store the resulting counter rate in',
+      }),
+      types: ['string'],
+      required: false,
+    },
+  },
+
+  fn(input, { by, inputColumnId, outputColumnId, outputColumnName }) {
+    const resultColumns = buildResultColumns(
+      input,
+      outputColumnId,
+      inputColumnId,
+      outputColumnName
+    );
+
+    if (!resultColumns) {
+      return input;
+    }
+    const previousValues: Partial<Record<string, number>> = {};
+    return {
+      ...input,
+      columns: resultColumns,
+      rows: input.rows.map((row) => {
+        const newRow = { ...row };
+
+        const bucketIdentifier = getBucketIdentifier(row, by);
+        const previousValue = previousValues[bucketIdentifier];
+        const currentValue = newRow[inputColumnId];
+        if (currentValue != null && previousValue != null) {
+          if (Number(currentValue) >= previousValue) {
```

Comment:
nit: There are four places where `currentValue` is cast to number, we could refactor this to a single `currentValueAsNumber`, but no biggie